evmrpc/filter.go

```diff
@@ -0,0 +1,199 @@
+package evmrpc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/filters"
+	abci "github.com/tendermint/tendermint/abci/types"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+)
+
+type filter struct {
+	fc      filters.FilterCriteria
+	cursors map[common.Address]string
+}
+
+type FilterAPI struct {
+	tmClient     rpcclient.Client
+	nextFilterID uint64
+	filters      map[uint64]filter
+}
+
+func NewFilterAPI(tmClient rpcclient.Client) *FilterAPI {
+	filters := make(map[uint64]filter)
+	return &FilterAPI{tmClient: tmClient, nextFilterID: 1, filters: filters}
+}
+
+func (a *FilterAPI) NewFilter(
+	ctx context.Context,
+	crit filters.FilterCriteria,
+) (*uint64, error) {
+	curFilterID := a.nextFilterID
+	a.nextFilterID++
+	a.filters[curFilterID] = filter{crit, make(map[common.Address]string)}
+	return &curFilterID, nil
+}
+
+func (a *FilterAPI) GetFilterChanges(
+	ctx context.Context,
+	filterID uint64,
+) ([]*ethtypes.Log, error) {
+	filter, ok := a.filters[filterID]
+	if !ok {
+		return nil, errors.New("filter does not exist")
+	}
+	res, cursors, err := a.getLogsOverAddresses(ctx, filter.fc, filter.cursors)
+	if err != nil {
+		return nil, err
+	}
+	updatedFilter := a.filters[filterID]
+	updatedFilter.cursors = cursors
+	a.filters[filterID] = updatedFilter
+	return res, nil
+}
+
+func (a *FilterAPI) GetFilterLogs(
+	ctx context.Context,
+	filterID uint64,
+) ([]*ethtypes.Log, error) {
+	filter, ok := a.filters[filterID]
+	if !ok {
+		return nil, errors.New("filter does not exist")
+	}
+	noCursors := make(map[common.Address]string)
+	res, cursors, err := a.getLogsOverAddresses(ctx, filter.fc, noCursors)
+	if err != nil {
+		return nil, err
+	}
+	updatedFilter := a.filters[filterID]
+	updatedFilter.cursors = cursors
+	a.filters[filterID] = updatedFilter
+	return res, nil
+}
+
+func (a *FilterAPI) GetLogs(
+	ctx context.Context,
+	crit filters.FilterCriteria,
+) ([]*ethtypes.Log, error) {
+	logs, _, err := a.getLogsOverAddresses(
+		ctx,
+		crit,
+		make(map[common.Address]string),
+	)
+	return logs, err
+}
+
+// pulls logs from tendermint client over multiple addresses.
+func (a *FilterAPI) getLogsOverAddresses(
+	ctx context.Context,
+	crit filters.FilterCriteria,
+	cursors map[common.Address]string,
+) ([]*ethtypes.Log, map[common.Address]string, error) {
+	res := make([]*ethtypes.Log, 0)
+	if len(crit.Addresses) == 0 {
+		crit.Addresses = append(crit.Addresses, common.Address{})
```

Comment:
what's the reason for appending an empty address here?

Reply:
If the user doesn't pass in an address, we need to search over all addresses. I basically want it to enter the for loop below but with an empty address. Let me know if you see a better way to do this. I tried to think about some other ways of doing this, but I think this might not be a bad way.